### PR TITLE
AKCORE-2: Initial interfaces for KafkaShareConsumer

### PR DESCRIPTION
**Pull request by AndrewJSchofield**
Defined external interfaces for KafkaShareConsumer following the pattern of KafkaConsumer. Not functional yet but will give other team members an interface to build against.

There are no tests at this point.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/ShareConsumer.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * @see KafkaShareConsumer#acknowledge(ConsumerRecord, AcknowledgeType)
+     */
+    void acknowledge(ConsumerRecord<K, V> record, AcknowledgeType type);
+
+    /**
+     * @see KafkaShareConsumer#commitSync()
+     */
+    Map<TopicIdPartition, Optional<KafkaException>> commitSync();
```
Comment (apoorvmittal10) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ShareConsumer.java:73]: Though it's debatable and not sure if wrapping the exception here (as method params) would be beneficial as with Optional as well we need a null check regarding if optional is not null.

**Review by apoorvmittal10 — APPROVED** [review]
LGTM

File: clients/src/main/java/org/apache/kafka/clients/consumer/AcknowledgeCommitCallback.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+     * @param results A map of the results for each topic-partition for which delivery was acknowledged.
+     *                If the acknowledgement failed for a topic-partition, an exception is present.
+     *
+     * @throws WakeupException if {@link KafkaShareConsumer#wakeup()} is called before or while this
+     *             function is called
+     * @throws InterruptException if the calling thread is interrupted before or while
+     *             this function is called
+     * @throws KafkaException for any other unrecoverable errors (e.g. if offset metadata
+     *             is too large or if the committed offset is invalid).
+     */
```
[the remark this answers is not in the document]
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/AcknowledgeCommitCallback.java:46]: Yes, the KIP needs to do this. The list will be longer than this and it needs to be in the KIP before voting.

File: clients/src/main/java/org/apache/kafka/clients/consumer/ShareConsumer.java (hunk 1 above)
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/ShareConsumer.java:73]: It will contain `Optional.empty()` if there's no exception.

Comment (apoorvmittal10) [reply clients/src/main/java/org/apache/kafka/clients/consumer/ShareConsumer.java:73]: Sure, just thinking if Optional<KafkaException> vs KafkaException as param, though I know client code has Optional params and I added similar for KIP-714 but sometimes I think it's unnecessary wrapping.

**Pull request merged**